test/test-manager/src/tests/account.rs

```diff
@@ -207,6 +208,82 @@ pub async fn test_revoked_device(
     Ok(())
 }
 
+/// Assert that an old Wireguard key is automatically rotated by the daemon.
+#[test_function]
+pub async fn test_automatic_wireguard_rotation(
+    ctx: TestContext,
+    rpc: ServiceClient,
+    mut mullvad_client: MullvadProxyClient,
+) -> anyhow::Result<()> {
+    const ROTATION_TIMEOUT: Duration = Duration::from_secs(120);
+
+    // Make note of current WG key
+    let old_key = get_current_wireguard_key(&mut mullvad_client).await?;
+
+    log::info!("Old wireguard key: {old_key}");
+
+    rpc.stop_mullvad_daemon().await?;
+
+    log::info!("Changing created field of `device.json` to more than 7 days ago");
+    rpc.make_device_json_old()
+        .await
+        .context("Could not change device.json to have an old created timestamp")?;
+
+    rpc.start_mullvad_daemon().await?;
+
+    // NOTE: Need to create a new `mullvad_client` here after the restart otherwise we can't
+    // communicate with the daemon
+    log::info!("Reconnecting to daemon");
+    drop(mullvad_client);
+    let mut mullvad_client = ctx.rpc_provider.new_client().await;
+
+    log::info!("Verifying that wireguard key has changed");
+    let event_listener = mullvad_client
+        .events_listen()
+        .await
+        .context("Failed to begin listening for state changes")?;
+    let new_key = get_current_wireguard_key(&mut mullvad_client).await?;
+
+    // If key has not yet been updated, listen for changes to it
+    if new_key == old_key {
+        // Verify rotation has happened within `ROTATION_TIMEOUT` - if the key hasn't been rotated after
+        // that, the rotation probably won't happen anytime soon.
+        log::info!("Listening for device daemon event");
+        let device_event = |daemon_event| match daemon_event {
+            DaemonEvent::Device(device_event) => Some(device_event),
+            _ => None,
+        };
+        let device_event_listener = tokio::time::timeout(
+            ROTATION_TIMEOUT,
+            helpers::find_daemon_event(event_listener, device_event),
+        );
+        let _ = device_event_listener.await;
+
+        // Note: The key rotation could possible have happened without us noticing due to
+        // some raceiness in the timeframe between starting the daemon and us starting to
+        // listen for new daemon events. Thus, it is probably a good idea to check manually if the
+        // device key was rotated.
```

Comment:
I think I like my version of this comment better (used to be above the creation of the event listener). Was there something about it you disagreed with?